gateway/gateway.go

```diff
@@ -258,28 +298,32 @@ func (gw *Gateway) handleMessage(msg config.Message, dest *bridge.Bridge) []*BrM
 			}
 		}
 		flog.Debugf("=> Sending %#v from %s (%s) to %s (%s)", msg, msg.Account, originchannel, dest.Account, channel.Name)
+
 		msg.Channel = channel.Name
 		msg.Avatar = gw.modifyAvatar(origmsg, dest)
 		msg.Username = gw.modifyUsername(origmsg, dest)
 		msg.ID = ""
-		if res, ok := gw.Messages.Get(origmsg.ID); ok {
-			IDs := res.([]*BrMsgID)
-			for _, id := range IDs {
-				// check protocol, bridge name and channelname
-				// for people that reuse the same bridge multiple times. see #342
-				if dest.Protocol == id.br.Protocol && dest.Name == id.br.Name && channel.ID == id.ChannelID {
-					msg.ID = id.ID
-				}
-			}
-		}
+		msg.ParentID = ""
+
+		msg.ID = gw.getDestMsgID(origmsg.ID, dest, channel)
+
 		// for api we need originchannel as channel
 		if dest.Protocol == "api" {
 			msg.Channel = originchannel
 		}
+
+		if canonicalParentMsgID != "" {
+			msg.ParentID =  gw.getDestMsgID(canonicalParentMsgID, dest, channel)
+			if msg.ParentID == "" {
+			   msg.ParentID = canonicalParentMsgID
+			}
+		}
+
```

Comment:
As in my suggestion above, msg.ParentID will be empty or have the correct id, so above can be changed to

```
if msg.ParentID == "" {
      msg.ParentID = canonicalParentMsgID
}
```